llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/mem0/base.py

```diff
@@ -0,0 +1,201 @@
+from typing import Dict, List, Optional, Union, Any
+from llama_index.core.memory.chat_memory_buffer import ChatMemoryBuffer
+from llama_index.core.memory.types import BaseMemory
+from llama_index.memory.mem0.utils import (
+    convert_memory_to_system_message,
+    convert_chat_history_to_dict,
+    convert_messages_to_string,
+)
+from mem0 import MemoryClient, Memory
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    ValidationError,
+    model_validator,
+    SerializeAsAny,
+    PrivateAttr,
+)
+from llama_index.core.base.llms.types import ChatMessage, MessageRole
+
+
+class BaseMem0(BaseMemory):
+    """Base class for Mem0."""
+
+    _client: Optional[Union[MemoryClient, Memory]] = PrivateAttr(default=None)
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def __init__(self, **data) -> None:
+        super().__init__(**data)
+        if "client" in data:
+            object.__setattr__(self, "_client", data["client"])
+
+    # TODO: Return type
+    def add(
+        self, messages: Union[str, List[Dict[str, str]]], **kwargs
+    ) -> Optional[Any]:
+        if self._client is None:
+            raise ValueError("Client is not initialized")
+        return self._client.add(messages=messages, **kwargs)
+
+    # TODO: Return type
+    def search(self, query: str, **kwargs) -> Optional[Any]:
+        if self._client is None:
+            raise ValueError("Client is not initialized")
+        return self._client.search(query=query, **kwargs)
+
+    # TODO: Add more apis from client
+
+
+class Mem0Context(BaseModel):
+    user_id: Optional[str] = None
+    agent_id: Optional[str] = None
+    run_id: Optional[str] = None
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    @model_validator(mode="after")
+    def check_at_least_one_assigned(cls, values):
+        if not any(
+            getattr(values, field) for field in ["user_id", "agent_id", "run_id"]
+        ):
+            raise ValueError(
+                "At least one of 'user_id', 'agent_id', or 'run_id' must be assigned."
+            )
+        return values
+
+    def get_context(self) -> Dict[str, Optional[str]]:
+        return {key: value for key, value in self.__dict__.items() if value is not None}
+
+
+class Mem0Memory(BaseMem0):
+    chat_history: SerializeAsAny[BaseMemory] = Field(
+        description="Primary memory source for chat agent."
+    )
+    _context: Optional[Mem0Context] = PrivateAttr(default=None)
+    search_msg_limit: int = Field(
+        default=5,
+        description="Limit of chat history messages to use for context in search API",
+    )
+
+    def __init__(self, **data) -> None:
+        super().__init__(**data)
+        if "context" in data:
+            object.__setattr__(self, "_context", data["context"])
```

Comment:
Similar to my comment above, if `context` is mandatory, let's make it mandatory.

If it is not mandatory, we can just type the constructor, instead of using **data

```
def __init__(self, context: Optional[Mem0Context] = None, ...):
  super().__init__()
  self._context = context
```

If it doesn't need to be private, we don't even need `__init__`